### PR TITLE
Attention mask as PackedSeqParams

### PR DESCRIPTION
**Pull request by jalbericiola**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/rl/rl_utils.py [hunk 1]
```diff
@@ -512,6 +513,59 @@ def pack_sequences(
...
+    Args:
+        packing_info: Dictionary containing packing metadata from SequencePacker
+        bin_idx: Index of the bin to create params for
+        bin_size: Size of the bin (padded sequence length)
+        device: Device to create tensors on
+
+    Returns:
+        PackedSeqParams with cu_seqlens set for proper attention masking
+    """
+    seq_indices = packing_info['bin_seq_indices'][bin_idx]
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:538]: I would really like to have packing_info as a dataclass and not just a dict. But given how urgent this fix is, let's add a todo?

File: megatron/rl/rl_utils.py [hunk 2]
```diff
@@ -512,6 +513,59 @@ def pack_sequences(
...
+    # Handle empty bins (padding bins with no sequences)
+    if not seq_indices:
+        return None
+
+    # Get actual sequence lengths for sequences in this bin
+    seq_lengths_in_bin = [packing_info['seq_lengths'][idx] for idx in seq_indices]
+
+    # Build cumulative sequence lengths for actual sequences
+    # cu_seqlens should be [0, len(seq1), len(seq1)+len(seq2), ..., total_actual_len]
+    cu_seqlens_list = [0]
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: nit: would be cleaner with np.cumsum([0] + seq_lengths_in_bin)?

File: megatron/rl/rl_utils.py [hunk 3]
```diff
@@ -512,6 +513,59 @@ def pack_sequences(
...
+    for length in seq_lengths_in_bin:
+        cumsum += length
+        cu_seqlens_list.append(cumsum)
+
+    cu_seqlens = torch.tensor(cu_seqlens_list, dtype=torch.int32, device=device)
+
+    max_seqlen = max(seq_lengths_in_bin) if seq_lengths_in_bin else bin_size
+
+    return PackedSeqParams(
+        qkv_format='thd',
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:556]: Is there any potential bug if someone decides to use a different qkv format? Or thd is pretty much hardcoded everywhere?

File: megatron/rl/rl_utils.py [hunk 4]
```diff
@@ -721,21 +779,53 @@ def get_logprobs(model, tokens, position_ids, attention_mask, no_grad=False):
...
+            if packed_seq_params is not None and packed_seq_params.qkv_format == 'thd':
+                # Get the actual token count (excluding padding)
+                actual_len = packed_seq_params.cu_seqlens_q[-1].item()
+                if actual_len == 0:
+                    # No real tokens, skip packed path
+                    packed_seq_params = None
+                else:
+                    # Slice inputs to remove padding
+                    tokens = tokens[:, :actual_len]
+                    position_ids = position_ids[:, :actual_len]
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:799]: Can you add a comment here explaining that batch dim is always 1 for packed, and that actual_len is the same for all (one) element.

File: megatron/rl/rl_utils.py [hunk 5]
```diff
@@ -1620,25 +1710,43 @@ def prepare_data_for_update(
...
                 else:
                     b_attn_mask = None
 
+                b_trajs = b_trajs.cuda()
+                b_posids = b_posids.cuda()
+
+                # Create packed_seq_params for proper attention masking in TE
+                b_packed_seq_params = None
+                if use_packed_computation and 'packing_info' in packing_context:
+                    bin_idx = batch_idx
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: nit: get rid of this variable and send batch_idx directly to create_packed_seq_params_for_bin

File: megatron/rl/rl_utils.py [hunk 6]
```diff
@@ -1708,16 +1816,31 @@ def prepare_data_for_update(
...
+                    start_idx = batch_idx * logprobs_batch_size
                     end_idx = min(
-                        start_idx + args.micro_batch_size, compute_attention_mask.shape[0]
+                        start_idx + logprobs_batch_size, compute_attention_mask.shape[0]
                     )
                     b_attn_mask = compute_attention_mask[start_idx:end_idx].cuda()
                 else:
                     b_attn_mask = None
 
+                b_trajs = b_trajs.cuda()
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:1843]: Let's wrap this into an inner function to avoid copypasting the same logic twice, please?

File: examples/rl/environments/math/math_agent.py [hunk 7]
```diff
@@ -21,8 +21,7 @@
     MATHVERIFY_AVAILABLE
 ), "math_verify is not installed but now required. Install it using `pip install math-verify` to continue."
 
-NEGATIVE_REWARD = 0.0
```
Comment (yobibyte) [inline examples/rl/environments/math/math_agent.py:24]: Should we remove the environment changes from this PR?

Comment (jalbericiola) [reply examples/rl/environments/math/math_agent.py:24]: yes, we should remove the changes in the math_agent from this PR

File: megatron/rl/rl_utils.py (hunk 5 above)
Comment (jalbericiola) [reply megatron/rl/rl_utils.py]: changed this in the new version with caching

File: megatron/rl/rl_utils.py (hunk 3 above)
Comment (jalbericiola) [reply megatron/rl/rl_utils.py:556]: in theory QKV format and the packed sequences params format should be orthogonal and handled by TE internally
